source/common/http/conn_manager_impl.cc

```diff
@@ -1801,19 +1801,20 @@ void ConnectionManagerImpl::ActiveStreamDecoderFilter::
 
 void ConnectionManagerImpl::ActiveStreamDecoderFilter::addDownstreamWatermarkCallbacks(
     DownstreamWatermarkCallbacks& watermark_callbacks) {
-  // This is called exactly once per stream, by the router filter.
```

Comment:
"Once per stream" could be updated for clarity.

Before, upstream and downstream streams functionally had 1:1 correspondence (well, to be really picky, one a time for retries).  Once per upstream-stream?

Also I'd be concerned that if you had an upstream which was timing out, it wouldn't be acking data, so your hedge would fail.  This is somewhat mitigated by the fact retries currently work for fully buffered requests, but could cause your hedged request to fail due to the outstanding stream, a problematic design flaw which might be worth documenting.  Alternately if we can do more to detect problematic configs, say one with a larger listener perConnectionBufferLimitBytes than cluster perConnectionBufferLimitBytes with hedging on.

Reply:
Ah interesting, I hadn't thought of that. Do you mind explaining a bit more why the case where listener buffer size is larger than cluster connection buffer size is particularly problematic? Not sure I get that part.